share/minizinc/std/distribute.mzn

```diff
@@ -7,8 +7,8 @@ include "fzn_distribute_reif.mzn";
 */
 predicate distribute(array[int] of var int: card,
                      array[int] of var int: value,
-                     array[int] of var int: base) =
+                     array[$X] of var int: base) =
     assert(index_set(card) == index_set(value),
```

Comment:
I'm not sure if this is possible using the current MiniZinc templating system, but a variant
```
predicate distribute(array[$Y] of var int: card,
                     array[$Y] of var int: value,
                     array[$X] of var int: base) =
```
with some way of asserting that the $Y variables have to represent the same dimensions would be even better.

Reply:
The same is of course applicable for all the global cardinality constraints.

Reply:
Actually I think this can be done using the `index_sets_agree` predicate:
```
predicate distribute(array[$Y] of var int: card,
                     array[$Y] of var int: value,
                     array[$X] of var int: base) =
    assert(index_sets_agree(card, value),
        "distribute: card and value arrays must have identical index sets",
        fzn_distribute(array1d(card), array1d(value), array1d(base))
    );
```
However, I am not completely sure, so I will leave this as an exercise for someone else. :)

Reply:
Yes, `index_sets_agree` is the right way to go. The `$Y` notation only makes sure that the static properties of the arrays are the same (i.e., same number of dimensions). The actual index sets cannot be checked by the type checker since they're only available at runtime, so the check has to be done explicitly.